Lift rpc_timeout to RpcAgent, for other RpcAgents to reuse.

xush6528 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29341 Lift rpc_timeout to RpcAgent, for other RpcAgents to reuse.**

https://github.com/pytorch/pytorch/pull/29336

So that other RpcAgent could use this timeout setting as well.

Differential Revision: [D5681951](https://our.internmc.facebook.com/intern/diff/D5681951/)

xush6528 committed:
Pass global_process_timeout to RpcAgent

https://github.com/pytorch/pytorch/pull/29336

- Rename the concept from "rpcTimeout" to "globalProcessTimeout", because 1) this timeout is used to limit the server-side request processing time. 2) This timeout is a global setting in contract to per-RPC setting.

Differential Revision: [D5681951](https://our.internmc.facebook.com/intern/diff/D5681951/)

[ghstack-poisoned]

xush6528 committed:
Update on "Pass global_process_timeout to RpcAgent"


https://github.com/pytorch/pytorch/pull/29336

- Rename the concept from "rpcTimeout" to "globalProcessTimeout", because 1) this timeout is used to limit the server-side request processing time. 2) This timeout is a global setting in contract to per-RPC setting.

Differential Revision: [D5681951](https://our.internmc.facebook.com/intern/diff/D5681951/)

[ghstack-poisoned]

xush6528 committed:
Update on "Pass global_process_timeout to RpcAgent"


https://github.com/pytorch/pytorch/pull/29336

- Rename the concept from "rpcTimeout" to "globalProcessTimeout", because 1) this timeout is used to limit the server-side request processing time. 2) This timeout is a global setting in contract to per-RPC setting.

Differential Revision: [D5681951](https://our.internmc.facebook.com/intern/diff/D5681951/)

[ghstack-poisoned]

xush6528 committed:
Update on "Pass global_process_timeout to RpcAgent"


https://github.com/pytorch/pytorch/pull/29336

- Rename the concept from "rpcTimeout" to "globalProcessTimeout", because 1) this timeout is used to limit the server-side request processing time. 2) This timeout is a global setting in contract to per-RPC setting.

Differential Revision: [D5681951](https://our.internmc.facebook.com/intern/diff/D5681951/)

[ghstack-poisoned]

pietern reviewed (commented):
@pritamdamania87 commented in the original PR https://github.com/pytorch/pytorch/pull/29336#issuecomment-550585781 about why there should be 3 timeouts and I agree. From the client's perspective there is no difference between the phases you list. Allowing different timeouts to be used means it's practically impossible to reason about timeouts (or spot them in log output) anymore.

Looking at the contents of the PR, I see two things:
1. Lifting `rpcTimeout` from ProcessGroupAgent into RPCAgent and renaming it
2. Unrelated formatting changes

We can all agree that lifting the parameter to RPCAgent is a good idea. Renaming is a separate issue and assumes we're going to have a bunch of timeouts, so we should discuss this more before going through with it. The formatting changes should be a separate formatting-only PR so it keeps the history usable (or you'll end up investigating why those lines were touched as part of a timeout change).

I propose the following:
1. Remove the rename from this PR and make it only move the timeout field to the base class.
2. Submit an issue for the rename so we can discuss what you're proposing.
3. Submit a separate PR for the formatting changes.

xush6528 commented:
@pietern For 2), opened #29402 to explain the whole context and idea.

For 1), 3), reverting in this PR.

xush6528 committed:
Update on "Pass global_process_timeout to RpcAgent"


https://github.com/pytorch/pytorch/pull/29336

- Rename the concept from "rpcTimeout" to "globalProcessTimeout", because 1) this timeout is used to limit the server-side request processing time. 2) This timeout is a global setting in contract to per-RPC setting.

Differential Revision: [D5681951](https://our.internmc.facebook.com/intern/diff/D5681951/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/rpc_agent.h:
```diff
@@ -93,6 +96,11 @@ class TORCH_API RpcAgent {
   virtual const WorkerInfo& getWorkerInfo(worker_id_t id) const = 0;
 
+  // Retrieve the timeout for all RPCs.
+  virtual const std::chrono::milliseconds& getRpcTimeout() const {
```
Why this needs to be virtual. In which situation a derived class would like to override this value?

Can we make it inline?

xush6528 replied:
Removing it.

Marking it as inline.

Turned out pybind11 does not require it to be declared in sub-classes. pybind11 is able to look up this member function in the base class.

mrshenli commented on torch/distributed/rpc/__init__.py:
```diff
@@ -53,7 +53,10 @@ def init_model_parallel(
             num_send_recv_threads(int): Number of threads for send/recv work.
-            rpc_timeout (datetime.timedelta): Timeout for RPCs. Defaults to 10 seconds.
+            rpc_timeout (datetime.timedelta):
+                Global timeout for server to process an RPC request.
```
Before we made the decision on whether we should have finer grain timeout (server, client, etc.), shall we avoid mentioning server here?

xush6528 replied:
Makes sense. Removing it for now.

mrshenli commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -42,8 +42,10 @@ PyObject* rpc_init(PyObject* /* unused */) {
-              &RpcAgent::sync,
+              "sync", &RpcAgent::sync, py::call_guard<py::gil_scoped_release>())
+          .def(
+              "_get_rpc_timeout",
```
Question, if the RpcAgent will handle timeout internally, why do we need to expose this API in Python and C++? Is this only for test and debugging?

xush6528 replied:
I think what @rohan-varma mentioned was since we allow users to set timeout, we need to have a getter to allow users to inspect this value as well.

mrshenli replied:
I see, makes sense

mrshenli reviewed (approved):
LGTM! Approve to unblock. Please make sure tests pass before landing.

xush6528 committed:
Update on "Lift rpc_timeout to RpcAgent, for other RpcAgents to reuse."


https://github.com/pytorch/pytorch/pull/29336

So that other RpcAgent could use this timeout setting as well.

Differential Revision: [D5681951](https://our.internmc.facebook.com/intern/diff/D5681951/)

[ghstack-poisoned]

pietern commented:
@xush6528 Thanks!

pritamdamania87 commented on torch/distributed/rpc/api.py:
```diff
@@ -78,6 +79,11 @@ def _init_rpc(
     # Initialize RPC.
+    if not isinstance(rpc_timeout, datetime.timedelta):
+        raise RuntimeError(
+            "`rpc_timeout` must be a `datetime.timedelta`."
```
nit: I don't think we should have backticks (`) in exception messages. This would be displayed on the user's terminal and there is no markdown there.

xush6528 replied:
@pritamdamania87 

I was not expecting a markdown format. But want to make clear that it's a python variable. Any other suggestion to make it distinct? or you think plain text is fine?

pritamdamania87 commented on test/rpc_test.py:
```diff
@@ -962,13 +962,11 @@ def test_call_method_on_rref(self):
-    @requires_process_group_agent("PROCESS_GROUP rpc backend specific test, skip")
     @dist_init
     def test_get_default_rpc_timeout(self):
         timeout = rpc.get_rpc_timeout()
```
See the discussion in https://github.com/pytorch/pytorch/pull/28926, we should directly add docstrings to the pybind method.

rohan-varma replied:
@pritamdamania87 currently, `ProcessGroupAgent` is not exposed in the docs since I assumed that this is a private class and there's no reason for the user to directly access it (we provide higher level APIs for everything that needs to be done). If we wanted to move this doc to pybind, we'd have to expose documentation for `ProcessGroupAgent`, since `get_rpc_timeout` is a method on that class.

xush6528 replied:
Created issue #29628 for tracking thihs.

pritamdamania87 replied:
Can we move this method from `RpcAgent` to `torch.distributed.rpc` and the implementation uses the default RpcAgent?
